pagespeed/envoy/envoy_url_async_fetcher_test.cc

```diff
@@ -0,0 +1,215 @@
+  bool success() const { return success_; }
+  bool IsDone() const {
+    ScopedMutex lock(mutex_);
+    return done_;
+  }
+
+  virtual void Reset() {
+    ScopedMutex lock(mutex_);
+    AsyncFetch::Reset();
+    done_ = false;
+    success_ = false;
+    response_headers()->Clear();
+  }
+
+private:
+  AbstractMutex* mutex_;
+
+  GoogleString buffer_;
+  bool success_;
+  bool done_;
+
+  DISALLOW_COPY_AND_ASSIGN(EnvoyTestFetch);
+};
+
+} // namespace
+
+class EnvoyUrlAsyncFetcherTest : public ::testing::Test {
+public:
+  static void SetUpTestCase() {}
+
+protected:
+  EnvoyUrlAsyncFetcherTest()
+      : thread_system_(Platform::CreateThreadSystem()),
+        message_handler_(thread_system_->NewMutex()), flaky_retries_(0),
+        fetcher_timeout_ms_(FetcherTimeoutMs()) {}
+
+  virtual void SetUp() { SetUpWithProxy(""); }
+
+  static int64 FetcherTimeoutMs() {
+    return RunningOnValgrind() ? kFetcherTimeoutValgrindMs : kFetcherTimeoutMs;
+  }
+
+  void SetUpWithProxy(const char* proxy) {
+    const char* env_host = getenv("PAGESPEED_TEST_HOST");
+    if (env_host != NULL) {
+      test_host_ = env_host;
+    }
+    if (test_host_.empty()) {
+      test_host_ = kFetchHost;
+    }
+    GoogleString fetch_test_domain = StrCat("//", test_host_);
+    timer_.reset(Platform::CreateTimer());
+    statistics_.reset(new SimpleStats(thread_system_.get()));
+    EnvoyUrlAsyncFetcher::InitStats(statistics_.get());
+    envoy_url_async_fetcher_.reset(
+        new EnvoyUrlAsyncFetcher(proxy, thread_system_.get(), statistics_.get(), timer_.get(),
+                                 fetcher_timeout_ms_, &message_handler_));
+    mutex_.reset(thread_system_->NewMutex());
+
+    // Set initial timestamp so we don't roll-over monitoring stats right after
+    // start.
+    statistics_->GetUpDownCounter(EnvoyStats::kEnvoyFetchLastCheckTimestampMs)
+        ->Set(timer_->NowMs());
+  }
+
+  virtual void TearDown() {
+    // Need to free the fetcher before destroy the pool.
+    delete envoy_fetch_;
+    envoy_url_async_fetcher_.reset(NULL);
+    timer_.reset(NULL);
+    STLDeleteElements(&fetches_);
+  }
+
+  // Adds a new URL & expected response to the url/response structure, returning
+  // its index so it can be passed to StartFetch/StartFetches etc.
+  int AddTestUrl(const GoogleString& url, const GoogleString& content_start) {
+    urls_.push_back(url);
+    content_starts_.push_back(content_start);
+    int index = fetches_.size();
+    fetches_.push_back(new EnvoyTestFetch(
+        RequestContext::NewTestRequestContext(thread_system_.get()), mutex_.get()));
+    return index;
+  }
+
+  void StartFetch(int idx) {
+    fetches_[idx]->Reset();
+    envoy_url_async_fetcher_->Fetch(urls_[idx], &message_handler_, fetches_[idx]);
+  }
+
+  void StartFetches(size_t first, size_t last) {
+    for (size_t idx = first; idx <= last; ++idx) {
+      StartFetch(idx);
+    }
+  }
+
+  std::vector<EnvoyTestFetch*> fetches_;
```

Comment:
Shouldn't these be private?

Reply:
yes, moved to private.
code pushed in https://github.com/apache/incubator-pagespeed-mod/pull/1969/commits/8a92117ec74a99b4a1dffff479a71ca1b5db7717